Add support for axis relative keywords to for <position>

weinig opened the pull request:
#### 24929a7d9dcc0eaa8d8e9cdea6e5b2371bfe2d73
<pre>
Add support for axis relative keywords to for &lt;position&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=292069">https://bugs.webkit.org/show_bug.cgi?id=292069</a>

Reviewed by Darin Adler.

Adds support for the axis relative keywords for &lt;position&gt;
and &lt;bg-position&gt; including:

  - x-start
  - x-end
  - y-start
  - y-end

These are all resolved at style building time, so only the CSS specified
value representation, parsing and style conversion needed updating.

Enabled behind a preference flag CSSAxisRelativePositionKeywordsEnabled.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-x-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-x-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-y-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-y-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-valid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-computed.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-computed.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-invalid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-invalid.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-valid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-valid.tentative.html: Added.
    - Add and update tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new preference flag CSSAxisRelativePositionKeywordsEnabled that the new functionality is behind.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Pipe the preferences flag to the parser context for use by the parser.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
    - Support parsing the new forms.

* Source/WebCore/css/values/primitives/CSSPosition.h:
    - Support representing the new forms.

* Source/WebCore/style/values/primitives/StylePosition.cpp:
    - Support resolving the new forms.

Canonical link: <a href="https://commits.webkit.org/294158@main">https://commits.webkit.org/294158@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1596ff4538a6bb863bad174dbe82a8f226305f0d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100979 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10944 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106125 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51603 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103020 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29135 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76903 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33929 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103986 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16120 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57251 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15935 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9232 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50953 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93646 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85841 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9292 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108481 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/99588 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28107 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20708 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85871 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28469 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87409 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85411 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21738 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30126 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7853 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22158 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28037 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33305 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/123213 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27849 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34325 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31169 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29407 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-computed.tentative.html:
```diff
@@ -0,0 +1,188 @@
+        { "x-start": "left" },
+        { "x-end": "right" },
+        { "y-start": "top" },
+        { "y-end": "bottom" },
```
I think these are backwards?

weinig replied:
Why is that? My (perhaps admittedly flawed) understanding of sideways-lr was that it was just like vertical-lr except with the typographic mode set to horizontal, which I didn't think would effect this.

weinig replied:
When switching to use writingMode.isAnyTopToBottom(), this did indeed switch. Though I am still not sure I understand exactly why that is correct.

fantasai replied:
You just need to write a testcase (a layout one, not a JS one) and then it'll be obvious. :)

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-valid.tentative.html:
```diff
@@ -0,0 +1,53 @@
+test_valid_value(property, "y-end", "center y-end");
+test_valid_value(property, "10px y-end");
+test_valid_value(property, "left 10px y-end 20px");
+test_valid_value(property, "y-end 20px left 10px", "left 10px y-end 20px");
```
Would it make sense to add some percentage tests here also? (Can just flip some of these randomly from px to %.)

fantasai commented on Source/WebCore/style/values/primitives/StylePosition.cpp:
```diff
@@ -38,6 +38,83 @@ namespace Style {
 using namespace CSS::Literals;
 
+enum XAxisPhysicalKeywords : bool { Left, Right };
+enum YAxisPhysicalKeywords : bool { Top, Bottom };
```
Do we need to introduce these enums, or can we re-use `BoxSides`?

weinig replied:
These avoid the need to have invalid branches, since these functions can only return a subset.

weinig replied:
That said, I removed the functions and their uses entirely :).

fantasai commented on Source/WebCore/style/values/primitives/StylePosition.cpp:
```diff
@@ -38,6 +38,83 @@ namespace Style {
+        } else {
+            // vertical / right-to-left: `x-start` maps to `right`
+            return XAxisPhysicalKeywords::Right;
+        }
```
This should just use `writingMode.isAnyLeftToRight()`

fantasai commented on Source/WebCore/style/values/primitives/StylePosition.cpp:
```diff
@@ -38,6 +38,83 @@ namespace Style {
+            // vertical / right-to-left: `x-end` maps to `left`
+            return XAxisPhysicalKeywords::Left;
+        }
+    }
```
Ditto.

fantasai commented on Source/WebCore/style/values/primitives/StylePosition.cpp:
```diff
@@ -38,6 +38,83 @@ namespace Style {
+    } else {
+        // vertical: `y-start` maps to `top`
+        return YAxisPhysicalKeywords::Top;
+    }
```
This should use `writingMode.IsAnyTopToBottom()`.

fantasai commented on Source/WebCore/style/values/primitives/StylePosition.cpp:
```diff
@@ -38,6 +38,83 @@ namespace Style {
+    } else {
+        // vertical: `y-end` maps to `bottom`
+        return YAxisPhysicalKeywords::Bottom;
+    }
```
Ditto.

fantasai commented on Source/WebCore/style/values/primitives/StylePosition.cpp:
```diff
@@ -85,6 +162,76 @@ static auto resolveKeyword(CSS::Keyword::Center, const BuilderState&) -> LengthP
     return 50_css_percentage;
 }
 
+// MARK: Mapped value resolution
```
There's a lot of indirection here, and is it really necessary?

weinig replied:
I found it to be the most clear way of implementing it without duplicating a bunch of logic. From a performance perspective, it all inlines and compiles down to an optimal single set of branches. (Also, when the block/inline keywords are involved, they use the same logic.

weinig replied:
Actually, with the use of simpler writing mode check, I can reduce the cognitive overhead here a bit.

fantasai reviewed (commented):
r=fantasai
You'll have to get an accredited reviewer to sign off though. :)

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-x-computed-expected.txt:
```diff
@@ -2,8 +2,8 @@
 PASS Property background-position-x value 'right'
-FAIL Property background-position-x value 'x-start' assert_true: 'x-start' is a supported value for background-position-x. expected true got false
-FAIL Property background-position-x value 'x-end' assert_true: 'x-end' is a supported value for background-position-x. expected true got false
+FAIL Property background-position-x value 'x-start' assert_equals: expected "x-start" but got "0%"
+FAIL Property background-position-x value 'x-end' assert_equals: expected "x-end" but got "100%"
```
Do you understand what makes this, and the other case like it, fail?

weinig replied:
Yeah, the spec currently wants the computed value of the longhand to be the keyword, but the computed value of the shorthand to be the computed value. 

I don’t think having them differ makes sense, and I am discussing it.

weinig committed:
Add support for axis relative keywords to for <position>
https://bugs.webkit.org/show_bug.cgi?id=292069

Reviewed by Darin Adler.

Adds support for the axis relative keywords for <position>
and <bg-position> including:

  - x-start
  - x-end
  - y-start
  - y-end

These are all resolved at style building time, so only the CSS specified
value representation, parsing and style conversion needed updating.

Enabled behind a preference flag CSSAxisRelativePositionKeywordsEnabled.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-x-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-x-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-y-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/background-position-y-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/shape-function-valid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-computed.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-computed.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-invalid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-invalid.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-valid.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/position/position-valid.tentative.html: Added.
    - Add and update tests.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new preference flag CSSAxisRelativePositionKeywordsEnabled that the new functionality is behind.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Pipe the preferences flag to the parser context for use by the parser.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Position.cpp:
    - Support parsing the new forms.

* Source/WebCore/css/values/primitives/CSSPosition.h:
    - Support representing the new forms.

* Source/WebCore/style/values/primitives/StylePosition.cpp:
    - Support resolving the new forms.

Canonical link: https://commits.webkit.org/294158@main
